### PR TITLE
Feat: Crabby deposit

### PR DESCRIPTION
**Pull request by haythem96**
# Task: Crabby deposit

## Description
**Should be reviewed/merge after #673**

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

Please describe how to test to verify the changes. Provide instructions so we can reproduce.

## FE Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] Added video recordings if it is a UI change

## User Facing Checklist

- [ ] I fully understand the user problem this PR is solving
- [ ] I know who the target user is for this PR and have a deep understanding of that user
- [ ] I have tried this flow thinking from the pov of the target user for this PR
- [ ] I (or working w someone on team) have scheduled a user test for this PR (if it is a large change)

### CODE REVIEW

File: packages/bull-vault/src/BullStrategy.sol [hunk 1]
```diff
@@ -17,83 +24,111 @@ contract BullStrategy is ERC20, LeverageBull, UniBull {
     using StrategyMath for uint256;
 
     uint256 private constant ONE = 1e18;
+    uint32 private constant TWAP = 420;
```
Comment (daryakaviani) [inline packages/bull-vault/src/BullStrategy.sol:27]: Are we sure we want to stick with this TWAP? CC @aleone

File: packages/bull-vault/src/FlashBull.sol [hunk 2]
```diff
@@ -18,17 +18,16 @@ contract FlashBull {
 
     /// @dev enum to differentiate between Uniswap swap callback function source
     enum FLASH_SOURCE {
+        GENERAL_SWAP,
```
Comment (daryakaviani) [inline packages/bull-vault/src/FlashBull.sol:21]: What is this going to be used for?

**Review by daryakaviani — APPROVED** [review]
All the math matches the spec, unsure about `GENERAL_SWAP` callback and `TWAP` length though.

File: packages/bull-vault/src/FlashBull.sol (hunk 2 above)
Comment (haythem96) [reply packages/bull-vault/src/FlashBull.sol:21]: Selling borrowed USDC for ETH, added in this PR but didn't use it

File: packages/bull-vault/src/BullStrategy.sol (hunk 1 above)
Comment (aleone) [reply packages/bull-vault/src/BullStrategy.sol:27]: i think twap mostly won't matter for this contract, only initial deposits, so not super critical

File: packages/bull-vault/src/BullStrategy.sol [hunk 3]
```diff
@@ -17,83 +24,111 @@ contract BullStrategy is ERC20, LeverageBull, UniBull {
...
-    /// @dev the cap in ETH for the strategy, above which deposits will be rejected
-    uint256 public strategyCap;
+    /// @dev PowerToken controller
+    address public immutable powerTokenController;
 
     /// @dev true if Bull was initialized
     bool public isInitialized;
-
+    /// @dev the cap in ETH for the strategy, above which deposits will be rejected
+    uint256 public strategyCap;
```
Comment (aleone) [inline packages/bull-vault/src/BullStrategy.sol:46]: we might want to cap this in units of crab? could be eth as well and check the total leverage balance?

**Comment by aleone** [comment]
some comments

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 4]
```diff
@@ -1,43 +1,179 @@
...
+        uint256 usdcBorrowedBefore = IEulerDToken(dToken).balanceOf(address(bullStrategy));
+        uint256 userUsdcBalanceBefore = IERC20(usdc).balanceOf(user1);
+
+        vm.startPrank(user1);
+        IERC20(crabV2).approve(address(bullStrategy), _crabAmount);
+        bullStrategy.deposit{value: wethToLend}(_crabAmount);
+        vm.stopPrank();
+        
+        assertEq(bullStrategy.balanceOf(user1).sub(userBullBalanceBefore), bullToMint);
+        assertTrue(wethToLend.sub(IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy)).sub(ethInLendingBefore)) <= 2);
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:115]: why 2 here?

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 5]
```diff
@@ -1,43 +1,179 @@
...
+        assertEq(IEulerDToken(dToken).balanceOf(address(bullStrategy)).sub(usdcBorrowedBefore), usdcToBorrow);
+        assertEq(IERC20(usdc).balanceOf(user1).sub(userUsdcBalanceBefore), usdcToBorrow);
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:117]: should probably have some more assertions - crab received is what you expect, crab withdrawn is what you expect

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 6]
```diff
@@ -1,43 +1,179 @@
...
+        VaultLib.Vault memory strategyVault = IController(address(controller)).vaults(crabV2.vaultId());
+
+        return (strategyVault.collateralAmount, strategyVault.shortAmount);
+    }
+
+    function _calcBullToMint(uint256 _crabToDeposit) internal view returns (uint256) {
+        if (IERC20(bullStrategy).totalSupply() == 0) {
+            return _crabToDeposit;
+        } else {
+            uint256 share = _crabToDeposit.wdiv(IERC20(crabV2).balanceOf(address(bullStrategy)));
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:130]: this equation is not correct for calculating outside of the contract flow

in the contract, we deposit the _crabToDeposit to the contract before we do this calculation - in this case, we are doing it before we move the funds so we need to do:

```             uint256 share = _crabToDeposit.wdiv(IERC20(crabV2).balanceOf(address(bullStrategy)).add(_crabToDeposit)); ```

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 7]
```diff
@@ -1,43 +1,179 @@
...
+            uint256 bullTotalSupply = bullStrategy.totalSupply();
+            return share.wmul(bullTotalSupply).wdiv(uint256(1e18).sub(bullTotalSupply));
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:132]: needs to be updated to take into account the fix to make the denominator be (1-share) not 1-bullTotalSupply

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 8]
```diff
@@ -1,43 +1,179 @@
...
+                    uniBullHelper.getTwap(controller.wPowerPerpPool(), controller.wPowerPerp(), controller.weth(), TWAP, false);
+                (uint256 ethInCrab, uint256 squeethInCrab) = _getCrabVaultDetails();
+                uint256 crabUsdPrice = (ethInCrab.wmul(ethUsdPrice).sub(squeethInCrab.wmul(squeethEthPrice).wmul(ethUsdPrice)))
+                    .wdiv(crabV2.totalSupply());
+                wethToLend = bullStrategy.TARGET_CR().wmul(_crabToDeposit).wmul(crabUsdPrice).wdiv(ethUsdPrice);
+                usdcToBorrow = wethToLend.wmul(ethUsdPrice).wdiv(bullStrategy.TARGET_CR()).div(1e12);
+            }
+        }
+        else {
+            uint256 share = _crabToDeposit.wdiv(IERC20(crabV2).balanceOf(address(bullStrategy)));
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:154]: this is wrong for same reason as above (need to add _crabDeposit to the balance before calculating the share)

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 9]
```diff
@@ -1,43 +1,179 @@
...
+            wethToLend = IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy)).wmul(share).wdiv(uint256(1e18).sub(share));
+            usdcToBorrow = IEulerDToken(dToken).balanceOf(address(bullStrategy)).wmul(share).wdiv(uint256(1e18).sub(share)).div(1e12);
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:156]: no magic numbers por favor

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 10]
```diff
@@ -1,43 +1,179 @@
...
+        }
+
+        return (wethToLend, usdcToBorrow);
+    }
+
+    function _calcTotalEthDelta(uint256 _crabToDeposit) internal view returns (uint256) {
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:162]: I don't understand what this function does, and not sure its correct for whatever its trying to do.  Hard to know as don't see it called anywhere yet.

**Review by aleone — COMMENTED** [review]
some more test comments

File: packages/bull-vault/src/BullStrategy.sol (hunk 1 above)
Comment (haythem96) [reply packages/bull-vault/src/BullStrategy.sol:27]: This is removed

File: packages/bull-vault/src/BullStrategy.sol (hunk 3 above)
Comment (haythem96) [reply packages/bull-vault/src/BullStrategy.sol:46]: Yes, will add cap check in a future PR

File: packages/bull-vault/src/BullStrategy.sol [hunk 11]
```diff
@@ -17,83 +24,111 @@ contract BullStrategy is ERC20, LeverageBull, UniBull {
...
         uint256 bullToMint = _crabAmount;
 
         if (totalSupply() == 0) {
-            _mint(_from, _crabAmount);
+            _mint(msg.sender, _crabAmount);
         } else {
-            share = _crabAmount.wdiv(ERC20(crab).balanceOf(address(this)));
+            share = _crabAmount.wdiv(IERC20(crab).balanceOf(address(this)));
             uint256 bullTotalSupply = totalSupply();
             bullToMint = share.wmul(bullTotalSupply).wdiv(ONE.sub(bullTotalSupply));
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/src/BullStrategy.sol:99]: This is fixed

File: packages/bull-vault/src/BullStrategy.sol [hunk 12]
```diff
@@ -17,83 +24,111 @@ contract BullStrategy is ERC20, LeverageBull, UniBull {
...
-            _mint(_from, bullToMint);
+            _mint(msg.sender, bullToMint);
         }
 
-        // TODO: get prices
-        uint256 crabUsdPrice = 0;
-        uint256 ethUsdPrice = 0;
+        uint256 ethUsdPrice = _getTwap(ethUSDCPool, weth, usdc, TWAP, false);
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/src/BullStrategy.sol:103]: Done

File: packages/bull-vault/src/BullStrategy.sol [hunk 13]
```diff
@@ -17,83 +24,111 @@ contract BullStrategy is ERC20, LeverageBull, UniBull {
...
+        uint256 squeethEthPrice = _getTwap(ethWSqueethPool, wPowerPerp, weth, TWAP, false);
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/src/BullStrategy.sol:104]: Done

File: packages/bull-vault/src/LeverageBull.sol [hunk 14]
```diff
@@ -15,33 +22,82 @@ contract LeverageBull {
...
-    /// @dev USD debt in leverage component
-    uint256 public usdcDebt;
+    /// @dev USDC address
+    address internal usdc;
+    /// @dev WETH address
+    address internal weth;
+    /// @dev euler markets module
+    address internal eulerMarkets;
+    /// @dev euler eToken that represent the collateral asset
+    address internal eToken;
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/src/LeverageBull.sol:32]: Agree, added a task for it

File: packages/bull-vault/src/LeverageBull.sol [hunk 15]
```diff
@@ -15,33 +22,82 @@ contract LeverageBull {
...
         internal
+        returns (uint256, uint256)
     {
         uint256 ethToLend;
         uint256 usdcToBorrow;
 
         if (_bullShare == ONE) {
             ethToLend = TARGET_CR.wmul(_crabAmount).wmul(_crabPrice).wdiv(_ethUsdPrice);
-            usdcToBorrow = ethToLend.wmul(_ethUsdPrice).wdiv(TARGET_CR);
+            usdcToBorrow = ethToLend.wmul(_ethUsdPrice).wdiv(TARGET_CR).div(1e12);
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/src/LeverageBull.sol:71]: We have task for this

File: packages/bull-vault/src/BullStrategy.sol [hunk 16]
```diff
@@ -17,83 +24,111 @@ contract BullStrategy is ERC20, LeverageBull, UniBull {
...
         if (totalSupply() == 0) {
-            _mint(_from, _crabAmount);
+            _mint(msg.sender, _crabAmount);
         } else {
-            share = _crabAmount.wdiv(ERC20(crab).balanceOf(address(this)));
+            share = _crabAmount.wdiv(IERC20(crab).balanceOf(address(this)));
             uint256 bullTotalSupply = totalSupply();
             bullToMint = share.wmul(bullTotalSupply).wdiv(ONE.sub(bullTotalSupply));
-            _mint(_from, bullToMint);
+            _mint(msg.sender, bullToMint);
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/src/BullStrategy.sol:100]: fixed

File: packages/bull-vault/src/BullStrategy.sol [hunk 17]
```diff
@@ -17,83 +24,111 @@ contract BullStrategy is ERC20, LeverageBull, UniBull {
...
+        uint256 squeethEthPrice = _getTwap(ethWSqueethPool, wPowerPerp, weth, TWAP, false);
+        (uint256 ethInCrab, uint256 squeethInCrab) = _getCrabVaultDetails();
+        uint256 crabUsdPrice = (ethInCrab.wmul(ethUsdPrice).sub(squeethInCrab.wmul(squeethEthPrice).wmul(ethUsdPrice)))
+            .wdiv(IERC20(crab).totalSupply());
 
-        _leverageDeposit(bullToMint, share, crabUsdPrice, ethUsdPrice);
+        (, uint256 usdcBorrowed) = _leverageDeposit(bullToMint, share, crabUsdPrice, ethUsdPrice);
 
-        // TODO: sell USDC for ETH
+        IERC20(usdc).transfer(msg.sender, usdcBorrowed);
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/src/BullStrategy.sol:111]: nice catch, added task for it

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 18]
```diff
@@ -1,43 +1,179 @@
...
+    address internal usdc;
+    address internal euler;
+    address internal eulerMarketsModule;
+    address internal eToken;
+    address internal dToken;
+    address internal wPowerPerp;
 
-//         user1Pk = vm.envUint("USER_1_PK");
-//         user1 = vm.addr(user1Pk);
+    function setUp() public {
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:46]: this resolved

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 19]
```diff
@@ -1,43 +1,179 @@
...
+        (uint256 wethToLend, uint256 usdcToBorrow) = _calcCollateralAndBorrowAmount(crabToDeposit);
+
+        vm.startPrank(user1);
+        IERC20(crabV2).approve(address(bullStrategy), crabToDeposit);
+        bullStrategy.deposit{value: wethToLend}(crabToDeposit);
+        vm.stopPrank();
+
+        assertEq(bullStrategy.balanceOf(user1), crabToDeposit);
+        assertEq(IEulerDToken(dToken).balanceOf(address(bullStrategy)), usdcToBorrow);
+        assertTrue(wethToLend.sub(IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy))) <= 1);
```
[the remark this answers is not in the document]
Comment (haythem96) [reply packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:93]: I think rounding issue, there was no new block for accrued interest...

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol (hunk 4 above)
Comment (haythem96) [reply packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:115]: In fuzzing, with bigger numbers as input, the diff is > 1

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol (hunk 6 above)
Comment (haythem96) [reply packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:130]: this is fixed already in other PRs

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol (hunk 7 above)
Comment (haythem96) [reply packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:132]: fixed

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol (hunk 8 above)
Comment (haythem96) [reply packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:154]: fixed already

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol (hunk 9 above)
Comment (haythem96) [reply packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:156]: we have task

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol (hunk 10 above)
Comment (haythem96) [reply packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:162]: Yeah I see it is not used, I thought when started writing test I may need a function to get the bull delta

**Pull request merged**